File: src/defensive-pass.test.js
import { test } from "node:test";
import assert from "node:assert/strict";
import { initial, step, simulationStep } from "./game.js";
import { defensivePass, defensivePunt, goalThrow } from "./defensive-pass.js";
const u = 22.4 / 576;
test("goal lob threshold includes the displayed vertical pose origin", () => {
  const s=initial(), p=s.players[7];
  Object.assign(p,{x:380*u,z:16*u,physicalSprite:0}); p.stats[4]=100;
  assert.equal(goalThrow(s,7,0).high,false, "200 pixels equals twice throw skill");
  p.physicalSprite=50;
  assert.equal(goalThrow(s,7,0).high,true, "one-pixel pose shift crosses threshold");
});
function fixture() {
  const s = initial(),
    d = Array(18).fill(1000);
  for (const p of s.players) p.stun = 100;
  Object.assign(s.players[0], { x: 0, z: 0, stun: 0 });
  Object.assign(s.players[1], { x: 20 * u, z: 0, stun: 0 });
  Object.assign(s.players[3], { x: 100 * u, z: 0, stun: 0 });
  d[1] = 20;
  d[3] = 100;
  s.ball.owner = 0;
  return [s, d];
}
test("keeper prefers advanced roles, then falls back and applies strict throw range", () => {
  const [s, d] = fixture();
  let p = defensivePass(s, 0, d);
  assert.equal(p.receiver, 3);
  assert.equal(p.high, true);
  s.players[3].stun = 1;
  p = defensivePass(s, 0, d);
  assert.equal(p.receiver, 1);
  assert.equal(p.high, false);
  d[1] = 200;
  assert.equal(defensivePass(s, 0, d).high, true);
  d[1] = 201;
  assert.equal(defensivePass(s, 0, d), null);
});
test("pass lanes reject blocking directions; electroball clears those exclusions", () => {
  const [s, d] = fixture();
  Object.assign(s.players[16], { x: 50 * u, z: 0, stun: 0 });
  d[16] = 50;
  assert.equal(defensivePass(s, 0, d), null);
  s.ball.charged = true;
  assert.equal(defensivePass(s, 0, d).receiver, 3);
  Object.assign(s.players[4], { x: 100 * u, z: 0, stun: 0 });
  d[4] = 100;
  assert.equal(defensivePass(s, 0, d).receiver, 4);
});

test("defensive punts use goal offsets, strict steering threshold and mirrored wall routes", () => {
  const [s] = fixture();
  let plan = defensivePunt(s, 0, 0);
  assert.equal(plan.key, 3);
  assert.equal(plan.z, -48 * u);
  assert.equal(plan.steer, -1);
  assert.equal(plan.high, true);
  assert.equal(defensivePunt(s, 0, 50).steer, 0);
  assert.equal(defensivePunt(s, 0, 64).z, 48 * u);
  Object.assign(s.players[16], { x: 50 * u, z: 0, stun: 0 });
  for (const charged of [false, true]) {
    s.ball.charged = charged;
    for (const period of [1, 2]) {
      s.period = period;
      const d = period === 1 ? 1 : -1;
      s.players[16].x = d * 50 * u;
      for (const random of [0, 16]) {
        plan = defensivePunt(s, 0, random);
        assert.equal(plan.z, (random ? 288 : -288) * u);
        assert.equal(plan.x, d * 288 * u);
        assert.equal(plan.key, d * 3 + (random ? 1 : -1));
      }
    }
  }
});

test("AI keeper without receivers winds up a high punt and retains release steering", () => {
  const s = initial();
  for (const p of s.players) p.stun = 100;
  Object.assign(s.players[0], { x: 0, z: 0, stun: 0, aiWait: 0 });
  Object.assign(s.ball, { x: 0, z: 0, owner: 0 });
  step(s, simulationStep, {}, [false, false]);
  assert.equal(s.players[0].throwMode, 3);
  assert.equal(s.ball.owner, 0);
  const steer = s.players[0].throwSteer;
  for (let n = 0; n < 4; n++) step(s, simulationStep, {}, [false, false]);
  assert.equal(s.ball.owner, -1);
  assert.equal(s.ball.flightKind, 2);
  assert.equal(s.ball.vz, steer * 4 * 25 * u);
});

File: server/defensive_pass_test.go
package main

import "testing"

func defensivePassFixture() (State, [18]int) {
	const u = 22.4 / 576
	s := initial()
	var d [18]int
	for i := range s.Players {
		s.Players[i].Stun = 100
		d[i] = 1000
	}
	s.Players[0].X, s.Players[0].Z, s.Players[0].Stun = 0, 0, 0
	s.Players[1].X, s.Players[1].Z, s.Players[1].Stun = 20*u, 0, 0
	s.Players[3].X, s.Players[3].Z, s.Players[3].Stun = 100*u, 0, 0
	d[1], d[3] = 20, 100
	s.Ball.Owner = 0
	return s, d
}
func TestDefensivePassRolesAndRange(t *testing.T) {
	s, d := defensivePassFixture()
	p := s.defensivePass(0, &d)
	if p == nil || p.receiver != 3 || !p.high {
		t.Fatal("role priority", p)
	}
	s.Players[3].Stun = 1
	p = s.defensivePass(0, &d)
	if p == nil || p.receiver != 1 || p.high {
		t.Fatal("defender fallback", p)
	}
	d[1] = 200
	p = s.defensivePass(0, &d)
	if p == nil || !p.high {
		t.Fatal("strict throw range")
	}
	d[1] = 201
	if s.defensivePass(0, &d) != nil {
		t.Fatal("intelligence range")
	}
}
func TestDefensivePassBlockedDirections(t *testing.T) {
	const u = 22.4 / 576
	s, d := defensivePassFixture()
	s.Players[16].X, s.Players[16].Z, s.Players[16].Stun = 50*u, 0, 0
	d[16] = 50
	if s.defensivePass(0, &d) != nil {
		t.Fatal("blocked lane")
	}
	s.Ball.Charged = true
	if p := s.defensivePass(0, &d); p == nil || p.receiver != 3 {
		t.Fatal("charged ball")
	}
	s.Players[4].X, s.Players[4].Z, s.Players[4].Stun = 100*u, 0, 0
	d[4] = 100
	if p := s.defensivePass(0, &d); p == nil || p.receiver != 4 {
		t.Fatal("equal-distance roster tie")
	}
}

func TestDefensivePunt(t *testing.T) {
	const u = 22.4 / 576
	s, _ := defensivePassFixture()
	p := s.defensivePunt(0, 0)
	if p.key != 3 || p.z != -48*u || p.steer != -1 || !p.high {
		t.Fatal("goal punt", p)
	}
	if s.defensivePunt(0, 50).steer != 0 || s.defensivePunt(0, 64).z != 48*u {
		t.Fatal("random bit or strict threshold")
	}
	s.Players[16].Stun, s.Players[16].Z = 0, 0
	for _, charged := range []bool{false, true} {
		s.Ball.Charged = charged
		for _, period := range []int{1, 2} {
			s.Period = period
			d := s.direction(0)
			s.Players[16].X = d * 50 * u
			for _, random := range []int{0, 16} {
				p = s.defensivePunt(0, random)
				lateral := -1.
				if random != 0 {
					lateral = 1
				}
				if p.z != lateral*288*u || p.x != d*288*u || p.key != int(d*3+lateral) {
					t.Fatal("wall route", charged, period, random, p)
				}
			}
		}
	}
}

func TestKeeperPuntRelease(t *testing.T) {
	s := initial()
	for i := range s.Players {
		s.Players[i].Stun = 100
	}
	p := &s.Players[0]
	p.X, p.Z, p.Stun, p.aiWait = 0, 0, 0, 0
	s.Ball.X, s.Ball.Z, s.Ball.Owner = 0, 0, 0
	s.simulate(simulationStep, [2]Input{}, [2]bool{})
	if p.throwMode != 3 || s.Ball.Owner != 0 {
		t.Fatal("missing high punt windup")
	}
	steer := p.throwSteer
	for n := 0; n < 4; n++ {
		s.simulate(simulationStep, [2]Input{}, [2]bool{})
	}
	if s.Ball.Owner != -1 || s.Ball.FlightKind != 2 || s.Ball.VZ != steer*4*velocityUnit {
		t.Fatal("punt release", s.Ball)
	}
}
